feat: add kube_deployment_status_replicas_terminating and kube_replicaset_status_terminating_replicas metrics

atiratree opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/guide/pull-requests.md#the-pull-request-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/pull-requests.md#write-release-notes-if-needed
4. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**:

This is needed to support the https://github.com/kubernetes/enhancements/issues/3973 observability and to promote DeploymentReplicaSetTerminatingReplicas to Beta https://github.com/kubernetes/kubernetes/pull/133087


**How does this change affect the cardinality of KSM**: *increases*

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

richabanker commented:
/triage accepted
/cc @rexagod

CatherineF-dev commented:
Error: ../../../go/pkg/mod/k8s.io/apiserver@v0.32.1/pkg/cel/environment/base.go:53:30: undefined: featuregate.DefaultComponentGlobalsRegistry
Error: ../../../go/pkg/mod/k8s.io/apiserver@v0.32.1/pkg/cel/environment/base.go:53:94: undefined: featuregate.DefaultKubeComponent
Error: ../../../go/pkg/mod/k8s.io/apiserver@v0.32.1/pkg/cel/environment/base.go:55:30: undefined: utilversion.DefaultKubeEffectiveVersion

Could you upgrade k8s version in a separate pull request?

atiratree commented:
I removed the go.mod bump requirement and rebased the PR.

Should be ready for a review @CatherineF-dev @richabanker @rexagod 

@soltysh for SIG Apps

atiratree commented on internal/store/deployment.go:
```diff
@@ -160,6 +160,26 @@ func deploymentMetricFamilies(allowAnnotationsList, allowLabelsList []string) []
 			}),
 		),
+		*generator.NewFamilyGeneratorWithStability(
+			"kube_deployment_status_replicas_terminating",
```
This was originally done to use the same format as the other deployment metrics. But, I agree that the suggested format is better. Updated.
